pkg/webhooks/clusterlogging/clusterlogging.go

```diff
@@ -0,0 +1,285 @@
+		hint:       "Set MaxAge to a value <= 7d",
+		upperBound: TimeUnit("7d"),
+	}
+	ok, ret := appValidator.checkPolicy(retentionPolicy.App)
+	if !ok {
+		return ret
+	}
+
+	infraValidator := retentionPolicyValidator{
+		name:       "infra",
+		hint:       "MaxAge must be 0d",
+		upperBound: TimeUnit("0d"),
+	}
+	ok, ret = infraValidator.checkPolicy(retentionPolicy.Infra)
+	if !ok {
+		return ret
+	}
+
+	auditValidator := retentionPolicyValidator{
+		name:       "audit",
+		hint:       "Set MaxAge to a value <= 7d",
+		upperBound: TimeUnit("7d"),
+	}
+	ok, ret = auditValidator.checkPolicy(retentionPolicy.Audit)
+	if !ok {
+		return ret
+	}
+
+	return ret
+}
+
+// renderClusterLogging decodes an *cl.ClusterLogging from the incoming request.
+// If the request includes an OldObject (from an update or deletion), it will be
+// preferred, otherwise, the Object will be preferred.
+func (s *ClusterloggingWebhook) renderClusterLogging(request admissionctl.Request) (*cl.ClusterLogging, error) {
+	decoder, err := admissionctl.NewDecoder(&s.s)
+	if err != nil {
+		return nil, err
+	}
+	clusterLogging := &cl.ClusterLogging{}
+	if len(request.OldObject.Raw) > 0 {
+		err = decoder.DecodeRaw(request.OldObject, clusterLogging)
+	} else {
+		err = decoder.DecodeRaw(request.Object, clusterLogging)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return clusterLogging, nil
+}
+
+func le(lhs TimeUnit, rhs TimeUnit) (bool, error) {
+	lhsNumber, lhsUnit, err := parseTimeUnit(lhs)
+	if err != nil {
+		return false, err
+	}
+
+	rhsNumber, rhsUnit, err := parseTimeUnit(rhs)
+	if err != nil {
+		return false, err
+	}
+
+	return convertToSeconds(lhsNumber, lhsUnit) <= convertToSeconds(rhsNumber, rhsUnit), nil
+}
+
+func convertToSeconds(number uint64, unit string) uint64 {
+	if unit == "y" {
+		//This is not correct but will suffice for this webhook, since we have no valid retention > 7d
+		number = number * 365
+		unit = "d"
+	}
+	if unit == "M" {
+		//This is not correct but will suffice for this webhook, since we have no valid retention > 7d
+		number = number * 31
+		unit = "d"
+	}
+	if unit == "w" {
+		number = number * 7
+		unit = "d"
+	}
+
+	if unit == "d" {
+		number = number * 24
+		unit = "h"
+	}
+
+	if unit == "h" || unit == "H" {
+		number = number * 60
+		unit = "m"
+	}
+
+	if unit == "m" {
+		number = number * 60
+	}
+
+	return number
+}
+
+func parseTimeUnit(value TimeUnit) (uint64, string, error) {
+	match := reTimeUnit.FindStringSubmatch(string(value))
+	if match == nil || len(match) < 2 {
+		return 0, "", fmt.Errorf("unable to parse timeunit '%s' for invalid timeunit", value)
+	}
+
+	n := match[1]
+	number, err := strconv.ParseUint(n, 10, 0)
+	if err != nil {
+		return 0, "", fmt.Errorf("unable to parse uint '%s' ", n)
+	}
+	unit := match[2]
+	return number, unit, nil
+}
+
+// SyncSetLabelSelector returns the label selector to use in the SyncSet.
+func (s *ClusterloggingWebhook) SyncSetLabelSelector() metav1.LabelSelector {
+	return metav1.LabelSelector{
```

Comment:
I wonder if there's a way to do a merge from the default with set with the MatxhExpressions.

Reply:
If we don't (because it's not possible), maybe a note here about why?

Reply:
done